### PR TITLE
64339-fix-basho-bench-stop-issue

### PR DESCRIPTION
**Pull request by cdlwjing**
Changed:
- Add a stop_worker method to check whether still have alive worker
- Spwan a procss to kick off stop_worker function when a worker stop

BugzId: 64339

@chewbranca @djschlegel2

### CODE REVIEW

**Comment by djschlegel2** [comment]
@cdlwjing @chewbranca

I posted some thoughts in the original ticket (https://cloudant.fogbugz.com/f/cases/64339/Resolve-basho_bench_status-can-not-stop-issue).  I don't think we want to have workers try to manage their own supervisor; probably better if we expand the conditions under which basho_bench_duration decides its time to stop to include some counter-based option.

**Comment by chewbranca** [comment]
Yeah I agree with @djschlegel2, let's move this logic into `basho_bench_duration`. We're already using `basho_bench_duration` as the primary exit check [1], let's add the additional logic there (although we might need to rename `basho_bench_duration` to something like `basho_bench_terminator` with some witty commentary). 

@cdlwjing I would take the logic you added for looking for active workers and make a new function in `basho_bench_worker_sup` like `active_workers() -> [X || X <- workers(), X =/= undefined].`. Then where you're currently calling the new stop_worker function, I would instead call a function like: `basho_bench_duration:worker_stopping(WorkerPid)` and in then have a function in `basho_bench_duration` like: `worker_stopping(WorkerPid) -> gen_server:cast(?MODULE, {worker_stopping, WorkerPid}), ok.` with the corresponding function `handle_cast({worker_stopping, WorkerPid}, State) -> case basho_bench_work_sup:active_workers() -- [WorkerPid] of [] -> log_early_exit(), {stop, {shutdown, normal}, State}; _ -> maybe_end({noreply, State}) end.`

[1] https://github.com/cloudant/basho_bench/blob/cloudant/src/basho_bench.erl#L125-L134

File: src/basho_bench_terminator.erl [hunk 1]
```diff
@@ -1,4 +1,10 @@
--module(basho_bench_duration).
+%% -------------------------------------------------------------------
+%% Check whether to exit besho_bench, exit conditions
```
Comment (chewbranca) [inline src/basho_bench_terminator.erl]: Typo here: `s/besho/basho/`.

File: src/basho_bench_terminator.erl [hunk 2]
```diff
@@ -1,4 +1,10 @@
...
+%%     1. run out of duration time
+%%     2. run out of operations even if the duration time had not up
+%% -------------------------------------------------------------------
+
+-module(basho_bench_terminator).
```
Comment (chewbranca) [inline src/basho_bench_terminator.erl]: This commit should be two commits: 1) doing the rename from basho_bench_duration to basho_bench_terminator; 2) adding the additional logic. We do not mix whitespace/formatting/renaming/refactoring commits with actual logic changes, and in general it would be good to separate out the two into different PRs.

File: src/basho_bench_terminator.erl [hunk 3]
```diff
@@ -60,6 +67,15 @@ handle_call(remaining, _From, State) ->
     maybe_end({reply, Remaining, State}).
 
 
+handle_cast({worker_stopping, WorkerPid}, State) -> 
+    case basho_bench_worker_sup:active_workers() -- [WorkerPid] of 
+        [] ->
+            lager:info("The application has stopped early!"),
```
Comment (chewbranca) [inline src/basho_bench_terminator.erl]: Let's use the same debug macros we're using elsewhere.

**Comment by chewbranca** [comment]
@cdlwjing overall the logic looks good here. Please separate out the renaming of the module throughout the codebase from the logic that is added, or move it to a standalone PR. Also, the commit message in 34234dc3d03082b0fb7d4ac118f30e36b404e348 tags me and @djschlegel2, please don't tag people in commits, it doesn't accomplish what you want. Just reply in the PR pinging folks for further review, but not in the commits themselves. Once you separate out the commit into two and cleanup the message you'll want to squash (aka interactive rebase and remove commits) those commits out. Here's a decent tutorial on doing so if you're not familiar: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by chewbranca** [comment]
@cdlwjing also, if you find yourself writing bulleted line items in commit messages describing everything the commit does, you should just make separate commits.

File: src/basho_bench_duration.erl [hunk 4]
```diff
@@ -67,8 +67,8 @@ handle_call(remaining, _From, State) ->
     maybe_end({reply, Remaining, State}).
 
 
-handle_cast({worker_stopping, WorkerPid}, State) -> 
-    case basho_bench_worker_sup:active_workers() -- [WorkerPid] of 
+handle_cast({worker_stopping, WorkerId}, State) ->
+    case basho_bench_worker_sup:active_workers() -- [WorkerId] of
```
Comment (cdlwjing) [inline src/basho_bench_duration.erl]: Find that the application can not be stopped in sometimes.
After looking into the basho_bench code, I found that the WorkerPid is the driver pid in actually(it is the pid of replication driver in this case, `WorkerPid = spawn_link(fun() -> worker_init(State) end), WorkerPid ! {init_driver, self()},`). But the pid list which returned from basho_bench_worker_sup:active_worker/1 is basho_bench_worker's.  So if driver has been stopped but worker hasn't been stopped very soon, then `basho_bench_worker_sup:active_workers() -- [WorkerPid]` will not return [], it will return worker process id.
Only if the worker has also been stopped very soon, then `[] -- [WorkerPid]` can still return []. That's  why the application can be stopped normally in sometimes and sometimes not.

Now I change it to `case basho_bench_worker_sup:active_workers() -- [WorkerId]`, the WorkerId is basho_bench_worker id, then no matter whether the worker stop very soon or not, then the app can stop normally.

Comment (chewbranca) [inline src/basho_bench_duration.erl]: Good catch. Let's switch the name here back to `WorkerPid` though, as the value passed in is still a pid, not an id.

**Comment by chewbranca** [comment]
@cdlwjing I'm +1 to this change once you've switched back to `WorkerPid` instead of `WorkerId`. Please squash the changes down to a single commit, force push back out to this branch, and then I'll do a final review before we merge.

**Comment by chewbranca** [comment]
+1

Going to merge this.
